nodeup/pkg/model/volumes.go

```diff
@@ -0,0 +1,151 @@
+type VolumesBuilder struct {
+	*NodeupModelContext
+}
+
+var _ fi.ModelBuilder = &VolumesBuilder{}
+
+// Build is responsible for handling the mounting additional volumes onto the instance
+func (b *VolumesBuilder) Build(c *fi.ModelBuilderContext) error {
+	// @step: check if the instancegroup has any volumes to mount
+	if !b.UseVolumeMounts() {
+		glog.V(1).Info("skipping the volume builder, no volumes defined for this instancegroup")
+
+		return nil
+	}
+
+	var mountUnits []string
+
+	// @step: iterate the volume mounts and create the format and mount units
+	for _, x := range b.InstanceGroup.Spec.VolumeMounts {
+		glog.V(2).Infof("attempting to provision device: %s, path: %s", x.Device, x.Path)
+
+		// @step: create the formatting unit
+		fsvc, err := buildDeviceFormatService(c, x)
+		if err != nil {
+			return fmt.Errorf("failed to provision format service for device: %s, error: %s", x.Device, err)
+		}
+		c.AddTask(fsvc)
+
+		msvc, err := buildDeviceMountService(c, x, fsvc.Name)
+		if err != nil {
+			return fmt.Errorf("failed to provision format service for device: %s, error: %s", x.Device, err)
+		}
+		c.AddTask(msvc)
+
+		mountUnits = append(mountUnits, msvc.Name)
+	}
+
+	// @step: create a unit for restart the docker daemon once everything is mounted
+	u := &systemd.Manifest{}
+	u.Set("Unit", "Description", "Used to start the docker daemon post volume mounts")
+	for _, x := range mountUnits {
+		u.Set("Unit", "After", x)
+		u.Set("Unit", "Requires", x)
+	}
+	u.Set("Service", "Type", "oneshot")
+	u.Set("Service", "RemainAfterExit", "yes")
+	u.Set("Service", "ExecStartPre", "/usr/bin/systemctl restart docker.service")
+	u.Set("Service", "ExecStart", "/usr/bin/systemctl restart --no-block kops-configuration.service")
+
+	c.AddTask(&nodetasks.Service{
+		Name:         b.EnsureSystemdSuffix(b.VolumesServiceName()),
+		Definition:   s(u.Render()),
+		Enabled:      fi.Bool(true),
+		ManageState:  fi.Bool(true),
+		Running:      fi.Bool(true),
+		SmartRestart: fi.Bool(true),
+	})
+
+	return nil
+}
+
+// buildDeviceFormatService is responsible for constructing the systemd unit to format the device
+func buildDeviceFormatService(c *fi.ModelBuilderContext, volume *kops.VolumeMountSpec) (*nodetasks.Service, error) {
+	device := volume.Device
+	deviceFmt := strings.TrimPrefix(strings.Replace(device, "/", "-", -1), "-")
+	name := fmt.Sprintf("format-%s.service", deviceFmt)
+
+	u := &systemd.Manifest{}
+	u.Set("Unit", "Description", fmt.Sprintf("Formats the device: %s", device))
+	u.Set("Unit", "After", fmt.Sprintf("%s.device", deviceFmt))
+	u.Set("Unit", "Requires", fmt.Sprintf("%s.device", deviceFmt))
+	u.Set("Service", "Type", "oneshot")
+	u.Set("Service", "RemainAfterExit", "yes")
+
+	// @TODO this was written to work on CoreOS need to check other OS's, add a switch on the distro and potentionally an api override
+	command := fmt.Sprintf("/usr/bin/bash -c '/usr/sbin/blkid %s || (/usr/sbin/wipefs -f %s && /usr/sbin/mkfs.%s %s)'",
```

Comment:
Yes, the consequence of this going wrong is so bad that I've always used SafeFormatAndMount from k/k

Reply:
Awesome shout!! .. i've switched over reusing the kubelet code to perform the mounting for me ...